Distribution/Server/Features/PackageInfoJSON/State.hs

```diff
@@ -0,0 +1,252 @@
+      getStatus  = (get :: Get Word8) >>= \case
+        0 -> return Preferred.NormalVersion
+        1 -> return Preferred.DeprecatedVersion
+        2 -> return Preferred.UnpreferredVersion
+        n -> fail $ "Unsupported tag for VersionStatus: " ++ show n
+
+
+-- | This encoding of @PackageVersions@ is used in the
+-- `/package/$package` endpoint (when the URI doesn't specify)
+-- a version. Any change here is an API change.
+instance Aeson.ToJSON PackageVersions where
+  toJSON (PackageVersions p) =
+    Aeson.toJSON
+    $ Map.mapKeys display
+    $ fmap encodeStatus
+    $ Map.fromList p
+    where
+      encodeStatus = \case
+        Preferred.NormalVersion      -> "normal"
+        Preferred.DeprecatedVersion  -> "deprecated"
+        Preferred.UnpreferredVersion -> "unpreferred"
+
+
+instance Aeson.FromJSON PackageVersions where
+  parseJSON = Aeson.withObject "PackageVersions" $ \obj ->
+    fmap PackageVersions
+    $ traverse (parsePair)
+    $ HashMap.toList obj
+    where
+      parsePair (vStr, vStatus) =
+        (,) <$> parseVersion vStr <*> parseStatus vStatus
+
+      parseVersion verText =
+        let verString = T.unpack verText
+        in case simpleParse verString of
+             Just ver -> return ver
+             Nothing  -> fail $ concat ["Could not parse \""
+                                       , verString ++ "\" as Version. "
+                                       , "expected \"a.b.c\" form"]
+
+      parseStatus (Aeson.String s) = case T.unpack s of
+        "normal"      -> return Preferred.NormalVersion
+        "deprecated"  -> return Preferred.DeprecatedVersion
+        "unpreferred" -> return Preferred.UnpreferredVersion
+        other         -> fail $ concat ["Could not parse \"" ++ other
+                         ++ "\" as status. Expected \"normal\""
+                         ++ "\"deprecated\" or \"unpreferred\""]
+      parseStatus _   = fail "Expected a string"
+
+data PackageInfoState = PackageInfoState {
+    descriptions          :: !(Map.Map (PackageIdentifier, Maybe Int) PackageBasicDescription)
+  , versions              :: !(Map.Map PackageName PackageVersions)
+  , migratedEphemeralData :: Bool
+  } deriving (Show, Typeable, Eq)
+
+getDescriptionFor
+  :: (PackageIdentifier, Maybe Int)
+  -> Query PackageInfoState (Maybe PackageBasicDescription)
+getDescriptionFor pkgId = asks $ Map.lookup pkgId . descriptions
+
+getVersionsFor
+  :: PackageName
+  -> Query PackageInfoState (Maybe PackageVersions)
+getVersionsFor pkgName = asks $ Map.lookup pkgName . versions
+
+setDescriptionFor
+  :: (PackageIdentifier, Maybe Int)
+  -> Maybe PackageBasicDescription
+  -> Update PackageInfoState ()
+setDescriptionFor  pkgId descr = State.modify $ \p ->
+  case descr of
+    Just d  -> p {descriptions = Map.alter (const (Just d)) pkgId (descriptions p)}
+    Nothing -> p {descriptions = Map.filterWithKey (\pkgId' _ -> fst pkgId' /= fst pkgId) (descriptions p)}
```

Comment:
@iamalsogreg can you explain why we wouldn't use Map.delete here?

Reply:
Yah - since `descriptions` is keyed on `(PackageIdentifier, Maybe Int)`, if I use `delete`, I would be deleting just a particular package/metadata combination. But what I want is for a change to the package to delete every package entry for some fixed package, at _every_ metadata revision.

If there were a `deleteWithKey` function that transforms the key before deciding what to delete, that could be nicer than the `filterWithKey` I use.